catalogs/catalog-fileset/src/main/java/org/apache/gravitino/catalog/fileset/SecureFilesetCatalogOperations.java

```diff
@@ -271,6 +280,211 @@ public void close() throws IOException {
+            if (entry != null) {
+              Object value = FieldUtils.readField(entry, "value", true);
+              if (value != null
+                  && value.getClass().getClassLoader() != null
+                  && value.getClass().getClassLoader()
+                      == SecureFilesetCatalogOperations.class.getClassLoader()) {
+                LOG.info(
+                    "Cleaning up thread local {} for thread {} with custom class loader",
+                    value,
+                    thread.getName());
+                FieldUtils.writeField(entry, "value", null, true);
+              }
+            }
+          }
+        }
+      } catch (Exception e) {
+        LOG.warn("Failed to clean up thread locals for thread {}", thread.getName(), e);
+      }
+    }
+  }
+
+  private static void closeStatsDataClearerInFileSystem() {
+    try {
+      ScheduledExecutorService scheduler =
+          (ScheduledExecutorService)
+              FieldUtils.readStaticField(MutableQuantiles.class, "scheduler", true);
+      scheduler.shutdownNow();
+      Field statisticsCleanerField =
+          FieldUtils.getField(FileSystem.Statistics.class, "STATS_DATA_CLEANER", true);
+      Object statisticsCleaner = statisticsCleanerField.get(null);
+      if (statisticsCleaner != null) {
+        ((Thread) statisticsCleaner).interrupt();
+        ((Thread) statisticsCleaner).setContextClassLoader(null);
+        ((Thread) statisticsCleaner).join();
+      }
+
+      FileSystem.closeAll();
+    } catch (Exception e) {
+      LOG.warn("Failed to close stats data clearer in FileSystem", e);
+    }
+  }
+
+  private static Thread[] getAllThreads() {
+    ThreadGroup rootGroup = Thread.currentThread().getThreadGroup();
+    ThreadGroup parentGroup;
+    while ((parentGroup = rootGroup.getParent()) != null) {
+      rootGroup = parentGroup;
+    }
+
+    Thread[] threads = new Thread[rootGroup.activeCount()];
+    while (rootGroup.enumerate(threads, true) == threads.length) {
+      threads = new Thread[threads.length * 2];
+    }
+    return threads;
+  }
+
+  private static boolean runningWithCurrentClassLoader(Thread thread) {
+    return thread != null
+        && thread.getContextClassLoader() == FilesetCatalogOperations.class.getClassLoader();
+  }
+
+  public static void clearShutdownHooks() {
+    try {
+      Class<?> shutdownHooks = Class.forName("java.lang.ApplicationShutdownHooks");
+      IdentityHashMap<Thread, Thread> hooks =
+          (IdentityHashMap<Thread, Thread>)
+              FieldUtils.readStaticField(shutdownHooks, "hooks", true);
+
+      hooks
+          .entrySet()
+          .removeIf(
+              entry -> {
+                Thread thread = entry.getKey();
+                return thread.getContextClassLoader()
+                    == FilesetCatalogOperations.class.getClassLoader();
+              });
+    } catch (Exception e) {
+      LOG.warn("Failed to clean shutdown hooks", e);
+    }
```

Comment:
The code above is very very brittle. You should separate this bunch of code from the main logic, and use another util class to encapsulate all the logic. That util class can be removed as a whole once we figure out a better way.

Another thing is that you should add lots of comments to describe each line of reflection, otherwise, others cannot understand why you made such a change.